Hw03 reflections

xRay64 opened the pull request.

Commit:
Homework 03 Java Professional Otus
Reflections and Annotations

Commit:
Merge branch 'main' into hw03-reflections

# Conflicts:
#	settings.gradle

Commit:
merge master

Commit:
Homework 03 Java Professional Otus
Reflections and Annotations
добавил проверку на наличие методов анатированных Before и After

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+
+public class TestClassParser {
+
+    private static Class testClass;
```
`Class<?>`

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+
+public class TestClassParser {
+
+    private static Class testClass;
+    private static Method beforeMethod;
+    private static Method afterMethod;
+    private static final List<Method> testMethodsList = new ArrayList<>();
+    private static int completeTestCount = 0;
+    private static int passTestCount = 0;
+    private static int failTestCount = 0;
```
Лучше не использовать статические поля для передачи данных между методами. Они про состояние класса, общее для всех его пользователей. В данном случае это явно не так. Плюс это плохо для запуска из нескольких потоков. 

Ну и если можно обойтись без состояния, то лучше обойтись. Эти поля могут быть локальными переменными, аргументами методов и возвращаемыми типами. Плюс некоторые из них можно сгруппировать в свои классы, типа контекста тестирования и результата тестирования. Тогда их будет удобнее передавать/возвращать

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+            try {
+                testClassObject = testClass.getDeclaredConstructor().newInstance();
+            } catch (Exception e) {
+                throw new RuntimeException("Тестовый клас должен иметь конструктор без аргументов");
```
Тут же любое исключение ловится? Вдруг в самом конструкторе без аргументов произойдет исключение?

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+        for (Method testMethod : testMethodsList) {
+
+            //Для каждого тестового метода создаем новый объект тестового класса
+            Object testClassObject;
+            try {
+                testClassObject = testClass.getDeclaredConstructor().newInstance();
+            } catch (Exception e) {
+                throw new RuntimeException("Тестовый клас должен иметь конструктор без аргументов");
+            }
```
Это можно вынести в метод. Возможно даже не этого класса

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+                try {
+                    afterMethod.invoke(testClassObject);
+                } catch (Exception e) {
+                    thrownException = e;
```
А мы так не перекроем возможное исключение в тестовом методе?

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+                }
+            }
+
+            //Сохраним результаты
+            completeTestCount++;
+            if (thrownException == null) {
+                passTestCount++;
+            } else {
+                failTestCount++;
+            }
```
Что-то тут возможно лишнее. У нас есть общее число тестов, это размер списка тестовых методов. Количество же успешных тестов равно общему количеству минус количество упавших

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,103 @@
+                if (beforeMethod == null) {
+                    beforeMethod = method;
+                } else {
+                    throw new RuntimeException("Тестовый класс может содержать только один метод Before");
```
Круто!

Commit:
Homework 03 Java Professional Otus
Reflections and Annotations
Добавил ReflectionHelper
Вынес контекст и результаты теста в отдельные классы
Изменил логику эксепшнов в тестовых методах

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,96 @@
+        TestContext testContext = new TestContext(clazz);
+        TestResult testResult = new TestResult();
+
+        parseClass(testContext);
```
Лучше `TestContext testContext = parseClass(clazz)`

stvort commented on hw03-testframework/src/main/java/ru/otus/testframework/processor/TestClassParser.java:
```diff
@@ -0,0 +1,96 @@
+                try {
+                    testContext.getBeforeMethod().invoke(testClassObject);
+                } catch (Exception e) {
+                    //если не смогли подготовить окружение - дальнешеее тестирование бессмысленно
```
Да. Но after надо выполнить. До исключения мы, например, могли открыть коннекшен к БД или создать файл, который за собой надо закрыть/удалить

Commit:
Merge branch 'main' into hw03-reflections

# Conflicts:
#	buildSrc/.gradle/7.2/executionHistory/executionHistory.bin
#	buildSrc/.gradle/7.2/executionHistory/executionHistory.lock
#	buildSrc/.gradle/buildOutputCleanup/buildOutputCleanup.lock
#	buildSrc/build/classes/java/main/Versions.class
#	buildSrc/build/libs/buildSrc.jar
#	buildSrc/build/tmp/compileJava/previous-compilation-data.bin

Commit:
Homework 03 Java Professional Otus
Reflections and Annotations
Переделал метод TestClassParser.parseClass
Добавил метод ReflectionHelper.callMethod
Добавил вызов after метода в случае эксепшна в before методе

Commit:
Homework 03 Java Professional Otus
Reflections and Annotations
Переделал логику обработки исключения в before методе
